R/io-csv.R

```diff
@@ -0,0 +1,284 @@
+#' @param credential_provider Provide a function that can be called to provide
+#' cloud storage credentials. The function is expected to return a dictionary
+#' of credential keys along with an optional credential expiry time.
+#' @param retries Number of retries if accessing a cloud instance fails.
+#' @param file_cache_ttl Amount of time to keep downloaded cloud files since
+#' their last access time, in seconds. Uses the `POLARS_FILE_CACHE_TTL`
+#' environment variable (which defaults to 1 hour) if not given.
+#' @param include_file_paths Include the path of the source file(s) as a column
+#' with this name.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' my_file <- tempfile()
+#' write.csv(iris, my_file)
+#' lazy_frame <- pl$scan_csv(my_file)
+#' lazy_frame$collect()
+#' unlink(my_file)
+pl__scan_csv <- function(
+    source,
+    ...,
+    has_header = TRUE,
+    separator = ",",
+    comment_prefix = NULL,
+    quote_char = '"',
+    skip_rows = 0,
+    schema = NULL,
+    schema_overrides = NULL,
+    null_values = NULL,
+    missing_utf8_is_empty_string = FALSE,
+    ignore_errors = FALSE,
+    cache = FALSE,
+    # with_column_names = NULL,
+    infer_schema = TRUE,
+    infer_schema_length = 100,
+    n_rows = NULL,
+    encoding = c("utf8", "utf8-lossy"),
+    low_memory = FALSE,
+    rechunk = FALSE,
+    skip_rows_after_header = 0,
+    row_index_name = NULL,
+    row_index_offset = 0,
+    try_parse_dates = FALSE,
+    eol_char = "\n",
+    # new_columns = NULL,
+    raise_if_empty = TRUE,
+    truncate_ragged_lines = FALSE,
+    decimal_comma = FALSE,
+    glob = TRUE,
+    storage_options = NULL,
+    # credential_provider = NULL,
+    retries = 2,
+    file_cache_ttl = NULL,
+    include_file_paths = NULL) {
+  wrap({
+    check_dots_empty0(...)
+    if (length(source) == 0) {
+      abort("`source` must have length > 0.")
+    }
+    check_list_of_polars_dtype(schema_overrides, allow_null = TRUE)
+    check_list_of_polars_dtype(schema, allow_null = TRUE)
+    check_named_list(schema_overrides, allow_null = TRUE)
+    check_named_list(schema, allow_null = TRUE)
+    check_named_list(storage_options, allow_null = TRUE)
+    encoding <- arg_match0(encoding, values = c("utf8", "utf8-lossy"))
+
+    if (isFALSE(infer_schema)) {
+      infer_schema_length <- 0
+    }
+
+    if (is.list(null_values)) {
+      all_character <- lapply(null_values, \(x) {
+        length(x) == 1 && is.character(x)
+      }) |>
+        unlist() |>
+        all()
+      if (!all_character) {
+        abort("When `null_values` is a list, each element must be of type character.")
+      }
+    } else if (is.character(null_values)) {
+      null_values <- as.list(null_values)
+    } else if (!is.null(null_values)) {
+      abort("`null_values` must be a character vector or a named list.")
+    }
+
+    schema_overrides <- lapply(schema_overrides, \(x) x$`_dt`)
+
+    if (length(schema) > 0) {
+      schema <- lapply(schema, \(x) x$`_dt`)
+    }
```

Comment:
Do you want to use `parse_into_list_of_datatypes`?

Reply:
Thanks, forgot about that one